### ISSUE TITLE
Original Model 128 worth it?

### ISSUE THREAD

**Issue by andenixa** [deepfakes/faceswap#385]
I have created a rough version of Original Model with dimensions 128, 128, 3.

Rationale:

There seems to be increasing demand for HD face-swapping while none had any luck with GAN128 as far as I can tell from issue and the playground. In addition it could also cover more face area.

Is releasing Original128 worth it?  Still assessing the efficiency. I had to sacrifice some color data to keep up with memory / speed limitations but overally its not very visible (as opposed to GAN128 that discards 
original color data). Speed seems to be up to snuff. Trying one-to-many scenarios as well. No LowMem version probably will ever be created for that.

Could also do Orig256 and Orig512 but definably won't fit in consumer GPU RAM.
--cheers

**Comment by Kirin-kun** [comment]
Post results eventually. 

Something like a short clip before and after. So we can judge if it's worth it?

I think the main problem isn't the resolution, but the averaging.

**Comment by andenixa** [comment]
Sure I will. I don't fully understand the theory though a highres data-set may contain more details for reconstruction (and more face coverage as well).

**Comment by ruah1984** [comment]
You can share out, and we run again to try.

**Comment by torzdf** [comment]
Yeah, I think it's worth it. If you can add a new model, then choice is good.

If it's in a state you can share, then please raise a Pull Request so others can test.

Thanks 👍

**Comment by ruah1984** [comment]
@andenixa we can test it, i believe 1080TI can support your request.

**Comment by andenixa** [comment]
@torzdf  yes its in in a state which I can share it. The major issue is to see if it has any meaningful results and since 128 models train much longer. I am still looking if it can perform (quality-wise) at the level of Original. At this stage it learns rather well but decoder part significantly lags behind the encoder and I can't predict its limitations.
Funny part I can run it with ENCODER_DIM of ~3k and batch size ~42 (there is no bs limitations) and it still fits 9080ti memory.

**Comment by andenixa** [comment]
Sure, I am yet to see consistent results that would at least outstretch GAN128. It gives me pretty good training ability, but I don't like the result it produces at this point. But I hope I am nearing that. Perhaps such nets need longer training.

**Comment by andenixa** [comment]
Still tuning the Net. Memory consumption is modest even for middle-level cards. Speed is quite good, but I can't get a crisp picture even for decoder.

**Comment by tjdwo13579** [comment]
I'm not trying to be nitpick but is conversion not possible with this model?

I've tried adding the "-t OriginalHighRes" on the conversion code but it's not working.

It says:
Failed to convert image: D:\Games\faceswap\Data\Video_in\00022.jpg. Reason: Error when checking : expected input_4 to have shape (None, 128, 128, 3) but got array with shape (1, 64, 64, 3)

Maybe the conversion script also needs to be changed?

**Comment by iperov** [comment]
check H128 on https://github.com/iperov/OpenDeepFaceSwap
this is like original half face model but for 128 and with DSSIM loss and separated mask trainer

**Comment by andenixa** [comment]
@tjdwo13579 you might be right I have forgotten to add conversion code. I did a PR yet I weren't able to test it with the latest git version. 
Somehow new releases became less Windows path friendly specially if you are using SMB paths like in my case.
@iperov I shall be sure to check it. Thanks.

**Comment by tjdwo13579** [comment]
@andenixa Thanks for adding the conversion code! I'll try it out now.

**Comment by andenixa** [comment]
@iperov I shall try your interleaved Upscale/ResBlock approach on decoder if you don't mind. I also like the face extractor you are using. I want to create something akin to H128 yet maskless. I noticed you reduce memory consumption by using smaller batch size. Does it play well for real data-sets? I noticed bugger batches contribute to more accurate / generalized models a lot. I wasn't able to refine anything with bs < ~45-48

**Comment by iperov** [comment]
any batch size <= 64  is fine.

https://www.youtube.com/watch?v=bprVuRxBA34 - this is trained on batch size = 1 on "lowmem" H64 2GB GPU

**Comment by andenixa** [comment]
@iperov looks excellent. Do you think its possible to preserve TARGET faces details, perhaps through another parallel layer? Such as freckles?

**Comment by iperov** [comment]
??

**Comment by iperov** [comment]
@andenixa model experiments with result comparison are much welcome.

**Comment by andenixa** [comment]
Thanks to @iperov I am currently testing another revision of HighRes model adopting his re-scaling idea.
Memory consumption is somewhat high though but you guys with 6 to 8Gb should be fine. Training speed is kind of slow though as there is much more deep layers now.

**Comment by torzdf** [comment]
I'll leave it open as a pr for now. Let me know when you think it's ready for merging.

**Comment by andenixa** [comment]
@torzdf sure, I am just trying to see if its not worse than the previous one considering decreased learning speed and raised memory demands. I shall credit the ideas I might have borrowed from other contributes of course. Generally I just want a working 128x tensor with HD quality;)

**Comment by andenixa** [comment]
Still working on the model. The clarity is fascinating now, but target vectors at some match wrongly aligned faces. I am trying to reduce number of deep layers to see if it helps that but I shall leave the reposit high clarity (very-deep) Encoder in the code for those who want to experiment.

**Comment by iperov** [comment]
@andenixa is SeparableConv useful ? what benefits it provides? have you comparison against regular idea with residual layers ?

**Comment by andenixa** [comment]
@iperov I think its slightly faster and less accurate with colors as it processes color layers separately (presumably sequentially). It generally consumes less memory though it probably has worse convergence in general. I need more layers to have a better generalization and I aim for better resolution for the first deep layer (ideally it has to be 2x the retina side size) yet I think it would be unfeasible to do with Conv2D (memory wise).

**Comment by torzdf** [comment]
Ok, I haven't got time to test this at the moment, but I will merge it into staging.

If anyone wants to checkout the staging branch, give it a go and report back their findings that would be appreciated.

**Comment by iperov** [comment]
@andenixa I make best H128, without suxx residual blocks. I removed res blocks from all models. New super update for all models upcoming...

**Comment by andenixa** [comment]
@iperov sounds fascinating. I shall check it out when I have time to make a proper training set for your Fork. Are H128 considerably different from full-face? For faceswap its just a matter of adjusting the margin matrix and of course training it to catch more "space".

PS: To achieve near production quality result I think we need something akin to GAN-network (not the GAN included with faceswap) that wound adjust learning rate dynamically, be incremental (adjust topology as needed) and have several pool and dropout layers.

**Comment by iperov** [comment]
H128 has more details vs full face, but doesnt cover one cheek and beard.
Half face good for women fakes which cheeks occluded by hair.

**Comment by andenixa** [comment]
@iperov I am not aiming to create fakes but rather to make one-to-many model, for a blog perhaps, (or actually for educational purposes) meaning I want to merge multiple faces in data set B to create a unique set of features. Like an avatar perhaps. I already have been successful with very basic Models by adding extra Conv layers and increasing Neuron count at dense layers. The problem of poor generalization and over-fitting still persists.

**Comment by iperov** [comment]
then what you doing in face swap repo ?

**Comment by andenixa** [comment]
@iperov faceswap it does work to some extent. It also doesn't have any working 128 model. I though I could provide any, but I am not sure if my "concoction" works good enough. Its has much better clarity now, but to be honest its very far from "HD" output.

**Comment by iperov** [comment]
@andenixa I am making avatar model too. Any thoughts about layers config ?

**Comment by iperov** [comment]
![h128_cage_0](https://user-images.githubusercontent.com/8076202/40107992-cd93358a-590a-11e8-92ca-a24aab3b3fda.jpg)

**Comment by andenixa** [comment]
@iperov if you mean one-to-many I have several thoughts but it would need testing.

For the network it would need one more layer for better generalization and slightly more neurons on Dense layers since samples differ one from another more drastically it would need to be able to learn that.

Yet I would start from pre-processing the target samples from set B before they are fed to the net
- Increase random transformation slightly more
- Slightly de-saturate the samples to make these appear less colorful. That reduces net preference to only one type of face based on hue.
- Normalize the luminance / brightness to make them appear as taken in slightly more similar lighting conditions.
Not sure how Keras training optimizers are exactly managing its _learning rate_ but I found that reducing it slightly at higher epochs helps the convergence. I am not sure if Adam optimizer does it automatically.

**Comment by iperov** [comment]
I pushed big update of OpenDeepFaceSwap. Quality signiciantly increased.

about avatar model, I mean 64 half face matching 256 avatar. 
![00000](https://user-images.githubusercontent.com/8076202/40110907-493dcd24-5912-11e8-9066-a80ce844d33b.png)

**Comment by andenixa** [comment]
@iperov thanks for the heads-up. I shall check it out.

Guys we should seriously stop using presidents for our testing purposes before they pass some legislations against face swapping.

**Comment by andenixa** [comment]
@iperov but your approach generally works. I am not sure to what extent mask is really necessary but without a mask H256 could fit to GPU.

**Comment by iperov** [comment]
I dont think so. 256 = 128 x 4 times bigger.

also mask very important. Model predicts mask smoothly instead of using source frame mask which much jittering.

**Comment by andenixa** [comment]
@iperov 
Do you want to contribute your idea to faceswap repo?
I can adapt OriginalHighRes model to use the mask, but I can't release it without your permission as its pretty derivative on your code.

**Comment by iperov** [comment]
@andenixa
> Do you want to contribute your idea to faceswap repo?
no :) Before forking I was contributing to faceswap. It is time wasting.

**Comment by andenixa** [comment]
@iperov I see, but I still can use it for my personal use, right?

The thing with **faceswap** is that it has wider audience specially with the emergence of the GUI which is very important for "layman".

**Comment by iperov** [comment]
of course you can use it. Why opensourcing then? :)

My windows binary has convenient .bat files and manual in Russian, so I dont need GUI absolutely.

**Comment by andenixa** [comment]
@iperov but I can't release the model that uses your masked approach right? Just to be clear.

**Comment by iperov** [comment]
masked trainer idea I got from dfaker repo.

**Comment by iperov** [comment]
I trying to make avatar model last 8 hours, but no success.
![python_2018-05-16_17-37-48](https://user-images.githubusercontent.com/8076202/40120398-21c7ef72-5930-11e8-8c28-b30abe1f5774.jpg)

**Comment by andenixa** [comment]
@iperov what is an avatar model. How is it different from the regular one?

**Comment by iperov** [comment]
half face control avatar. Goal is B half face will control A avatar

**Comment by andenixa** [comment]
@iperov do you think if less bits are used for colors, perhaps 15 bits it could be manageable to fit 256 vectors? Or create 128->256 dfaker approach. Its faking nothing if the output is blurry with common video standards such as fullHD and 4k.

**Comment by torzdf** [comment]
@andenixa https://github.com/deepfakes/faceswap/blob/master/LICENSE

It's all opensource, just credit where possible.

**Comment by iperov** [comment]
@andenixa you mean grayscale to rgb ? of course. NN will work as colorizer using average color of samples.

64->128 works like imaginator which produces jittering.
Forget about 256 for 2 years :)

**Comment by andenixa** [comment]
@iperov when a wise man claims something is possible they are probably right. When they say otherwise they might be mistaken. Two years is an awfully long time to those whose lack of patience only rivaled by the lack of health. Thanks for the council nevertheless. Your models are superb.

PS: Actually it ain't much work to come up with some GUI with PyQT if you don't mind an additional dependency.

**Comment by iperov** [comment]
GUI creates one more window. So there will be total 3 windows : GUI, console, preview.
If hide console and embed it to GUI, then if you close GUI window, child processes will not be closed gracefully due to python limitations of OS manipulations.
Therefore better work without GUI for me.

bat files is fine enough
![explorer_2018-05-16_21-27-35](https://user-images.githubusercontent.com/8076202/40133141-f13a3a20-594f-11e8-95ec-ead1ba66fee5.jpg)

**Comment by Jasas9754** [comment]
@andenixa any progress?

**Comment by andenixa** [comment]
@Jasas9754 I did make additional releases up to v8 then I realized I am doing what others already had done long ago. In general plain, unmasked autoencoder model has surprisingly low convergence as well very high memory demands and is trained over weeks. I was surprised when I learned that @iperov fork 128x is trained over 3-4 hours ready to be deployed and has a variety of 128x nets readily available.

**Comment by Jasas9754** [comment]
@andenixa LIAEF128 model learns faster than any model. That's something to learn. But i'm not still sure about eyeball tracking. Is this because of low dense(256)? Well, maybe nobody knows. If you want to experiment with something, but you don't have enough time, give us any experimental code. We'll test it out. Maybe you'd better use your own fork for feedback too.

**Comment by andenixa** [comment]
@Jasas9754 with LIAEF128 you need to give it slightly more Dense, probably 512, and make Encoder to look like 128->384->768->1024. You also need a crystal crisp data-set with no blurry input or output.
In my experience it would learn eyes in that configuration given enough (~60k) epochs

**Comment by andenixa** [comment]
I can commit my latest concoction of OriginalHighRes model. But it doesn't learn as good as LIAEF128 or maybe it does over many days.

**Comment by andenixa** [comment]
Actually @iperov was generous to provide me with 256x Model. That is when I realized that increasing output matrix is memory consuming and redundant.
There is 2-3 potential steps needed to create an UHD face swap.
1) Train a regular full-face, perhaps with DF model.
2) (optional) Depending on the crispness you can optionally train a de-blur filter. Because you have a high-res training set at hand as de-blur only works efficently for the data it was trained on.
3) Train additional super-resolution x2 or x4 tensor depending on your demands. It could be even x8 (seriously). You might use some serious enhanced de-noising super resolution approaches, but I recommend just a regular sr x4 algorithm trained on both A and B data-sets.
Apply 2) and 3) to your result and use CUBIC sampling to fit it back to a face hull.
That should give you Hollywood quality face swapping videos.

**Comment by luiscosio** [comment]
@andenixa You are working on hollywood quality face swapping :)?

**Comment by andenixa** [comment]
@luiscosio no I am not, but I have we would be able to pull it together.

**Comment by andenixa** [comment]
@ruah1984 I have a good idea, I need slightly more time to study TF.
Perhaps it might be the best idea ever seen in face swap and alike.

**Comment by andenixa** [comment]
I had several working experiments with a great help of @iperov, however I still haven't figured how to seamlessly apply filters to the whole image.

**Comment by ruah1984** [comment]
@andenixa looking for your new idea. You can apply here for try.

**Comment by tjess78** [comment]
hmn, latest push still says OOM in both modes, even with batch size 2

**Comment by Kirin-kun** [comment]
@tjess78 I think you have to checkout the andenixa-patch-1 branch

**Comment by Jasas9754** [comment]
It works fine. No doubt.

**Comment by andenixa** [comment]
I tested it. Seems to be learning rather well both performance and quality-wise.
Further increasing of image size would require a mask otherwise it won't converge because of the background.

**Comment by Jasas9754** [comment]
@andenixa 1.What are the benefits of 'shaoanlu' type? I don't think it's learning faster than the original, or clearer. What's the advantage?

2. You have raised encoder dim to 1024(not 512) and changed conv to conv_sep. Do you think encoder dim is more important?

**Comment by andenixa** [comment]
@Jasas9754 
1) I think it has better generalization, but you are not required to use it.
2. conv_sep takes less memory, but the final result is not different from conv, though it might be slower to catch up at later stages. Encoder dim is very important specially with unmasked auto-encoders. In fact its plays a great role for both learning, clarity, and better reconstruction. Though we don't always have the luxury to make it big enough because of video memory.

**Comment by tjess78** [comment]
Narf, I don't get it. No matter what I do, it will not run on my 1060 6GB. 
Here is the error. tried it with last push from gui 3.0 and andenixa patch 1,2 and 2-1. Could it be an error in my eviroment? 

Exception in thread Thread-1:
Traceback (most recent call last):
File "d:\ProgramData\Anaconda3\envs\gui\lib\threading.py", line 914, in _bootstrap_inner
self.run()
File "d:\ProgramData\Anaconda3\envs\gui\lib\threading.py", line 862, in run
self._target(*self._args, **self._kwargs)
File "D:\faceswap-and\scripts\train.py", line 97, in process_thread
raise err
File "D:\faceswap-and\scripts\train.py", line 89, in process_thread
self.run_training_cycle(model, trainer)
File "D:\faceswap-and\scripts\train.py", line 124, in run_training_cycle
trainer.train_one_step(epoch, viewer)
File "D:\faceswap-and\plugins\Model_OriginalHighRes\Trainer.py", line 39, in train_one_step
loss_B = self.model.autoencoder_B.train_on_batch(warped_B, target_B)
File "d:\ProgramData\Anaconda3\envs\gui\lib\site-packages\keras\engine\training.py", line 1220, in train_on_batch
outputs = self.train_function(ins)
File "d:\ProgramData\Anaconda3\envs\gui\lib\site-packages\keras\backend\tensorflow_backend.py", line 2661, in __call__
return self._call(inputs)
File "d:\ProgramData\Anaconda3\envs\gui\lib\site-packages\keras\backend\tensorflow_backend.py", line 2631, in _call
fetched = self._callable_fn(*array_vals)
File "d:\ProgramData\Anaconda3\envs\gui\lib\site-packages\tensorflow\python\client\session.py", line 1454, in __call__
self._session._session, self._handle, args, status, None)
File "d:\ProgramData\Anaconda3\envs\gui\lib\site-packages\tensorflow\python\framework\errors_impl.py", line 519, in __exit__
c_api.TF_GetCode(self.status.status))
tensorflow.python.framework.errors_impl.ResourceExhaustedError: OOM when allocating tensor with shape[65536,1024] and type float on /job:localhost/replica:0/task:0/device:GPU:0 by allocator GPU_0_bfc
[[Node: training_1/Adam/gradients/model_1/dense_1/MatMul_grad/MatMul_1 = MatMul[T=DT_FLOAT, _class=["loc:@training_1/Adam/gradients/model_1/dense_1/MatMul_grad/MatMul"], transpose_a=true, transpose_b=false, _device="/job:localhost/replica:0/task:0/device:GPU:0"](model_1/flatten_1/Reshape, training_1/Adam/gradients/model_1/dense_2/MatMul_grad/MatMul)]]
Hint: If you want to see a list of allocated tensors when OOM happens, add report_tensor_allocations_upon_oom to RunOptions for current allocation info.

**Comment by andenixa** [comment]
@tjess78 seems to be a genuine OOM. What batch size do you use and are there any other programs that use video memory are running in the background?
I generally test with a clean environment and a dedicated card. You might probably want to disable Aero theme is its enable and switch to basic. Also close any applications such as web browser that highly utilize video memory. You should be able to run it with batch sizes 8-13 (depending on free ram).

**Comment by Jasas9754** [comment]
People who are short on vram can adjust the settings.

```
ENCODER_DIM = 512

x = Dense(dense_shape * dense_shape * 1024)(x)
x = Reshape((dense_shape, dense_shape, 1024))(x)

x = self.upscale(512, kernel_initializer=RandomNormal(0, 0.02))(inpt)
x = self.upscale(256, kernel_initializer=RandomNormal(0, 0.02))(x)

```
I trained up to 100k with it(-bs 8). the quality is still better than the original 64. The speed is also acceptable.

**Comment by ruah1984** [comment]
can any one integrate  Iperov DeepFacelab to here? Because i think Deepfakes master here should include all kind of faceswap model study .  i not sure what will be the different but , i hope master here can include all different deepfakes model such as DFaker, original model, low memory model , LIAEF128YAW (5GB+)

**Comment by andenixa** [comment]
@ruah1984 I don't want do put you off as we might actually port some of these models eventually.
But if you are really interested in testing Iperov's models wouldn't it be better just to use his for for that purpose?

**Comment by ruah1984** [comment]
I have been try his work. And result looks good .Just hope if we have gui version and put all together inside here as a team.

**Comment by torzdf** [comment]
As @andenixa says. Anyone is welcome to open a PR to port other models.

**Comment by oracle9i88** [comment]
TypeError: join() argument  must be str or bytes, not 'PosixPath'

**Comment by oracle9i88** [comment]
other mode is fine

**Comment by tjess78** [comment]
**GAN v2.2 release**

2018-06-25 Model update: faceswap-GAN v2.2 has been released. The main improvements of v2.2 model are its capability of generating realistic and consistent eyeballs (results are shown below, or Ctrl+F for eyeballs), as well as its higher quality of video results with face alignment.

[https://github.com/shaoanlu/faceswap-GAN](Faceswap-GAN)

**Comment by agilebean** [comment]
Yes, I can confirm that the last reported bug (json expects str or bytes) is fixed.
OriginalHighRes works!
Thanks a lot to @andenixa and @torzdf for this extremely short turnaround time, really incredible.

**Comment by Kirin-kun** [comment]
I tested OriginalHighres with a small dataset and it looks really good. The faces are more detailed and have more depth/defined features than with Original.

A few differences: 

- OriginalHighRes seem to learn gaze direction better
- OriginalhighRes pupils are a little blurrier than Original and the color is lighter/grayish
- OriginalHighRes conversion seem to blend colors better. I had some patches of lighter skin on the cheeks for Original, but OriginalHighres gave better skin tone overall (same convert params)
- OriginalHighRes seem to have difficulties learning Open/Closed eyes and Open/Smiling mouth. I'm at 130k iterations and the smiles just start to look like the ones produced much faster by Original.

In then end, the only caveats are the pupils that look a little too like lifeless gray circles, in Original, they are more black so it's less visible. And eventually the smiles showing teeth that look blurred, but I will see if it improves after training more.

Overall, I'm really satisfied with this model. It might become my preferred model. And its memory management is amazing. With a GTX 1060 6gb, I can manage a batch size of 16

**Comment by andenixa** [comment]
@Kirin-kun thank you for your feedback. I think we essentially would need to add mask for eyes to work.
Batch size of `16` has nothing to do with that as it pretty sufficient. For now if you are doing something to produce a video, not just testing the model, I'd suggest to run the training to 300epochs. I know its an overkill but that would probably make the situation a little better with pupils and eye positions.

**Comment by Kirin-kun** [comment]
For the moment, I'm not doing videos. 

I use photos as source material. It's a lot better looking than flickering videos and it's possible to adjust convert parameters easier than with a video with thousands of frames with different zooming on the faces. I tried to do videos, but I had mixed results (with just about all models). In the end, still pictures of models posing gives the best visually seamless faceswap.

I trained more on the same dataset, from 130k to 150k iterations, and the changes are really minute on the converted faces. When comparing, the differences are barely visible.

Further improvements I see are, eventually, the pupils looking more lifelike and also a way to handle, at convert time, the obstacles, like glasses, hair, hand, hats, etc, that cover parts of the destination face.
